posteriors/laplace/dense_hessian.py

```diff
@@ -0,0 +1,144 @@
+    """Builds a transform for dense Hessian Laplace.
+
+    Args:
+        log_posterior: Function that takes parameters and input batch and
+            returns the log posterior value (which can be unnormalised)
+            as well as auxiliary information, e.g. from the model call.
+        init_prec: Initial precision matrix.
+            If it is a float, it is defined as an identity matrix
+            scaled by that float.
+
+    Returns:
+        Hessian Laplace transform instance.
+    """
+    init_fn = partial(init, init_prec=init_prec)
+    update_fn = partial(update, log_posterior=log_posterior)
+    return Transform(init_fn, update_fn)
+
+
+class DenseLaplaceState(NamedTuple):
+    """State encoding a Normal distribution over parameters,
+    with a dense precision matrix
+
+    Attributes:
+        params: Mean of the Normal distribution.
+        prec: Precision matrix of the Normal distribution.
+        aux: Auxiliary information from the log_posterior call.
+    """
+
+    params: TensorTree
+    prec: torch.Tensor
+    aux: Any = None
+
+
+def init(
+    params: TensorTree,
+    init_prec: torch.Tensor | float = 0.0,
+) -> DenseLaplaceState:
+    """Initialise Normal distribution over parameters
+    with a dense precision matrix.
+
+    Args:
+        params: Mean of the Normal distribution.
+        init_prec: Initial precision matrix.
+            If it is a float, it is defined as an identity matrix
+            scaled by that float.
+
+    Returns:
+        Initial DenseLaplaceState.
+    """
+
+    if is_scalar(init_prec):
+        num_params = tree_size(params)
+        init_prec = init_prec * torch.eye(num_params, requires_grad=False)
+
+    return DenseLaplaceState(params, init_prec)
+
+
+def update(
+    state: DenseLaplaceState,
+    batch: Any,
+    log_posterior: LogProbFn,
+    epsilon: float = 0.0,
+    rescale: float = 1.0,
+    inplace: bool = False,
+) -> DenseLaplaceState:
+    """Adds the Hessian of the negative log-posterior over given batch.
+
+    Args:
+        state: Current state.
+        batch: Input data to log_posterior.
+        log_posterior: Function that takes parameters and input batch and
+            returns the log posterior value (which can be unnormalised)
+        epsilon: Added to the diagonal of the Hessian
+            for numerical stability.
+        rescale: Value to multiply the Hessian by
+            (i.e. to normalize by batch size)
+        inplace: If True, the state is updated in place. Otherwise, a new
+            state is returned.
+
+    Returns:
+        Updated DenseLaplaceState.
+    """
+    with torch.no_grad(), CatchAuxError():
+        flat_params, params_unravel = tree_ravel(state.params)
+        num_params = flat_params.numel()
+
+        def neg_log_p(p_flat):
+            value, aux = log_posterior(params_unravel(p_flat), batch)
+            return -value, aux
+
+        hess, aux = jacfwd(jacrev(neg_log_p, has_aux=True), has_aux=True)(flat_params)
```

Comment:
This looks great, but just to be clear you've used `jacfwd(jacrev(` over `torch.func.hessian` because `torch.func.hessian` does not have the `has_aux=True` functionality?

Reply:
Yes, I used jacfwd and jacrev because hessian does not support has_aux=True.